Optimization of antichain inclusion

kilohsakul opened the pull request:
Worklist ordered by set size so small sets are processed first.

kilohsakul committed:
bullshit zmeny

kilohsakul committed:
Some polishing.

kilohsakul committed:
Some polishing.

kilohsakul committed:
Remove some static cast warnings.

kilohsakul committed:
Remove outdated comment.

Adda0 commented on src/nfa/inclusion.cc:
```diff
@@ -65,18 +65,27 @@ bool mata::nfa::algorithms::is_included_antichains(
+    auto smaller_set = [](const ProdStateType & a, const ProdStateType & b) { return a.second.size() < b.second.size(); };
+
+    //Inserting the pairs (q,S) into a sequence of pairs in order defined by the size of the sets S.
+    auto insert_to_pairs = [&](ProdStatesType & pairs,const ProdStateType & pair) {
+        auto it = std::lower_bound(pairs.begin(), pairs.end(), pair, smaller_set);
+        pairs.insert(it,pair);
+    };
```
This might be costly for `deque`. It might be worth investigating whether it can be done more efficiently.

kilohsakul replied:
Does not take time in the benchmark we were playing with. Also, google says that deque insert and erase are fast.

Adda0 replied:
Inserts are, but the lower bound might do something like iterating forward and backward time and again to find the position. Not sure about how it will interact with deque. That is what would need to be investigated.

Adda0 replied:
Replaced with `std::vector`, which is faster on a hard example: 4 seconds -> 3.6 seconds. We trade fast insert (in deque) for fast lower bound finding (in vector). We will see how the vector performs in our experiments.

Adda0 reviewed (approved):
Looks good to me. The performance increase is significant.

Adda0 committed:
Use vector instead of deque for worklist

We insert to the vector using iterators which is not ideal. But it seems
to work better than deque.
